db/migrate/20230719100256_migrate_recurring_count_and_unique_count_aggregation.rb

```diff
@@ -0,0 +1,151 @@
+# frozen_string_literal: true
+
+class MigrateRecurringCountAndUniqueCountAggregation < ActiveRecord::Migration[7.0]
+  # NOTE: redefine models to prevent schema issue in the future
+  class BillableMetric < ApplicationRecord; end
+  class Charge < ApplicationRecord; end
+  class Customer < ApplicationRecord; end
+  class Event < ApplicationRecord; end
+  class Plan < ApplicationRecord; end
+  class Subscription < ApplicationRecord; end
+
+  def change
+    reversible do |dir|
+      dir.up do
+        # Create quantified event object for all unique count events
+        sql = <<-SQL
+          SELECT
+            events.id AS event_id,
+            events.code AS code,
+            events.organization_id AS organization_id,
+            events.customer_id AS customer_id,
+            events.properties AS properties,
+            events.timestamp AS event_timestamp,
+            subscriptions.external_id AS subscription_external_id
+          FROM events
+          INNER JOIN subscriptions ON subscriptions.id = events.subscription_id
+          INNER JOIN billable_metrics ON billable_metrics.code = events.code
+          WHERE events.deleted_at IS NULL AND billable_metrics.aggregation_type = 3
+          ORDER BY event_timestamp ASC;
+        SQL
+
+        ApplicationRecord.connection.select_all(sql).each_with_object({}) do |row, _result|
+          matching_billable_metric = BillableMetric.find_by(
+            code: row['code'],
+            organization_id: row['organization_id'],
+          )
+
+          existing_quantified_event =
+            QuantifiedEvent.where(
+              customer_id: row['customer_id'],
+              billable_metric_id: matching_billable_metric.id,
+              external_subscription_id: row['subscription_external_id'],
+              external_id: JSON.parse(row['properties'])[matching_billable_metric.field_name.to_s],
+            ).where(removed_at: nil).any?
+
+          # There can only be one quantified event for certain external_id which guarantees uniqueness
+          next if existing_quantified_event
+
+          quantified_event = QuantifiedEvent.create!(
```

Comment:
Since here we cannot avoid using the model, I would suggest to redefine it in the migration, to prevent issue in the future if we change validations, attributes or anything else on the model.